### DISCUSSION TITLE
与Anki联动的最佳实践？

### DISCUSSION THREAD

**Discussion by makerjackie** [obsidianzh/forum#21] [实用技巧/最佳实践]
分享你的使用技巧，关于obsidian和Anki

**Comment by makerjackie** [comment]
# obsidian anki 教程
我们经常希望在 obsidian 记录的笔记能够方便地转换到 anki 中生成记忆卡片，具体要怎么做呢？

## 最终效果
假设你在obsidian的笔记长这样
```
- Question1
    - Answer

Question2::Answer

## Question3 #card
answer with 
multiline

## Not a question #card不在标题结尾，不会被转换
foo

\```
not a Question::should not convert 代码块中也不会被转换
\```

```
打开anki和obsidian，点击obsidian左侧菜单栏后
![image](https://user-images.githubusercontent.com/18050469/104087932-2096db80-529e-11eb-8e31-8f1d077782cb.png)

你的 anki 会多出如下卡片
![image](https://user-images.githubusercontent.com/18050469/104119080-2ce76b00-5368-11eb-88cb-62d5d64ef58a.png)



## 安装

1. anki 安装并配置 anki connector 插件
	* Open the `Install Add-on` dialog by selecting `Tools` | `Add-ons` | `Get Add-ons...` in Anki.
	* Input [2055492159](https://ankiweb.net/shared/info/2055492159) into the text box labeled `Code` and press the `OK` button to proceed.
	* Restart Anki when prompted to do so in order to complete the installation of AnkiConnect.
	* In Anki, navigate to Tools->Addons->AnkiConnect->Config, and change it to look like this:
```
	{
    "apiKey": null,
    "apiLogPath": null,
    "webBindAddress": "127.0.0.1",
    "webBindPort": 8765,
    "webCorsOrigin": "http://localhost",
    "webCorsOriginList": [
        "http://localhost",
        "app://obsidian.md"
    ]
}
```

2. obsidian 安装并配置 Obsidian_to_anki 插件
	* 在 `Community plugins` 中搜索并安装 `Obsidian_to_anki`，如果无法加载社区插件，请见[这里](https://github.com/obsidianzh/forum/discussions/25)
	* 在 `Third-party plugins` 中开启 `Obsidian_to_anki` 插件，重启 obsidian
	* 在 Plugin Options 中找到 obsidian_to_anki 的配置，在`Note Type Table` 的 `Basic` 卡 对应的 `Custom Regexp` 一栏中添加下面的内容
		*  `^\- ((?:.+\n)*?)\n*[\t ]+((\- )?.+(?:\n(?:^.{4}(?<!<!--)(?<!\- ..).*))*)(?:\n^$)?`
		* 这个"代码"[参考](https://github.com/Pseudonium/Obsidian_to_Anki/issues/192)
		
3. 你可以查看插件的[wiki英文帮助文档](https://github.com/Pseudonium/Obsidian_to_Anki/wiki)熟悉更多关于这个插件的配置


### 生成/更新卡片
配置完成后！随便找一个文档，输入上文中的（其他另外两种还需要翻到[#更多格式]一节进行设置）
```
- 问题1 
    - 答案
- 问题2 
    - 答案
```
然后生成/更新卡片，有两种方法
* 点击obsidian左侧栏中的anki图标，这可能要花一点时间，注意软件右上角的提示
* `ctrl + P` 呼出obsidian的命令框，输入`Obsidian_to_Anki: Scan Vault`，然后`Enter`执行即可。

## 常见问题
1. `-` 跟 markdown 的列表冲突了？
	1. 你可以用 `*` 来作为普通的列表，`* [ ]` 来标识 todo 
2. 如何删除卡片？
	1. `DELETE`加在` ID:xxxx ` 的上方
3. 如何避免符合上述格式的笔记被添加到 anki 中？
	1. 把内容包含在代码块里头，即在内容的前后添加一行  \`\`\`
4. obsidian 笔记有点多，如何折腾？
	1. `ctrl + p` ，然后在命令框中输入 `Fold all headings and lists` 折叠，`Unfold all headings and lists` 取消折叠
5. 和 flashcard 插件的区别？
* flashcard: 
	* 支持当前页笔记 导入到 anki（但不能一次扫描所有文件）
	* 支持两种格式的笔记（按下面的写法，就能够转换成 anki 卡片）
```
# Question #card 
Answer 

Question::Card # 这种写法 obsidian_to_anki 也支持，见下文的 `更多格式`
```
* Obsidian_to_anki
	* 支持扫描所有文件导入到 anki
	* 支持多种格式的笔记（需要多一步配置，如果懂正则表达式的话，还可以自己定义格式）
		* flashcard支持的格式它都支持
		* 填空题 `中间是一个{填空}题`
		* 其他格式（就不一一举例了，具体见: https://github.com/Pseudonium/Obsidian_to_Anki/wiki/Regex）
* 总之: 如果有上面自定义格式 + 扫描所有文件的刚需，可以用 obsidian_to_anki 插件（目前我在用的，我还写了一个自定义的笔记格式，见: https://github.com/Pseudonium/Obsidian_to_Anki/issues/192#issuecomment-756303155）
6. 有视频教程吗？
	* [obsidian_to_anki](https://github.com/Pseudonium/Obsidian_to_Anki) 
		* [视频教程-youtube](https://www.youtube.com/watch?v=nUAVySf-2Dk&list=PLWg9zacwOnwfcpVm5pAKgOHms7PntsgJS&index=1) 
	* [flashcard](https://github.com/reuseman/flashcards-obsidian/wiki) **推荐新手使用**
		* [视频教程-youtube](https://www.youtube.com/watch?v=tR0edWQakuU&list=PLWg9zacwOnwfcpVm5pAKgOHms7PntsgJS&index=2)
	* B站上应该有搬运的


### 更多格式
如果觉得上面的 笔记转anki格式 还不满足你的要求，例如你希望有挖空/填空题卡片，或者标题为问题，正文为内容，
你可以在这里查看所有 格式（regex）列表: https://github.com/Pseudonium/Obsidian_to_Anki/wiki/Regex

你需要做的事情就是复制链接中提供的 代码（一串字符） 到 插件设置里头的 `custom regex` 中

举例:
* 标题格式`^#+(.+)#card\n*((?:\n(?:^[^\n#].{0,2}$|^[^\n#].{3}(?<!<!--).*))+)`
* 双冒号格式 `^(.*[^\n:]{1}):{2}([^\n:]{1}.*)`

![Pasted image 20210110171058](https://user-images.githubusercontent.com/18050469/104119301-5d7bd480-5369-11eb-96e5-5ae0568ab17f.png)

几个例子
```
**RemNote single line style**

This is how to use::Remnote single-line style
The script won't see things outside of it.
You can have::multiple notes in the same file

**Cloze Paragraph style**

The idea of {cloze paragraph style} is to be able to recognise any paragraphs that contain {cloze deletions}.

The script should ignore paragraphs that have math formatting like $\\frac{3}{4}$ but no actual cloze deletions.

With {2:CurlyCloze} enabled, you can also use the {c1|easier cloze formatting},
but of course {{c3::Anki}}'s formatting is always an option.


**Question answer style**

Q: How do you use this style?
A: Just like this.

Q: Can the question
run over multiple lines?
A: Yes, and
So can the answer
```


### 已知问题
1. [css 中 有了 ID 后渲染有bug的问题](https://forum.obsidian.md/t/css-unorder-list-doesnt-show-correctly/11190)
2. Add context / tag 在 custom regex 中不可用

**Reply by Quorafind** [reply]
这个 reuseman 的项目，是已经可用了，但是比较简陋，无法自行选用卡片模板。

下边那个是另外的python脚本

**Reply by makerjackie** [reply]
在使用了，感觉第一个 flashcards-obsidian 用起来最为方便（不过现在有一些bug），第二个python脚本我看对普通用户操作上还有些麻烦（而且目前我还不是重度用户，就没有尝试了）

**Reply by Quorafind** [reply]
现在那个python脚本的作者自己写了个 Ob 插件；就叫做 Obsidian to anki

**Reply by makerjackie** [reply]
感谢，我尝试一下

**Comment by makerjackie** [comment]
up!

**Comment by lefinite** [comment]
可以的，非常方便。
